### PR TITLE
Telink b91 pinctrl

### PR DESCRIPTION
**Pull request by yurvyn**
This pull request contains the following changes:
- Major: Telink B91 new PINCTRL driver development (#39740).
- Minor: Telink B91 "fest flash" board documentation update and SPI_burn script improvement.

### CODE REVIEW

File: dts/bindings/pinctrl/telink,b91-pinctrl.yaml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      required: true
+
+    label:
+      required: true
+
+    pad-mul-sel:
+      type: int
+      required: true
+      description: |
+          PinMux pad_mul_sel register values. Pins functions depend on this value.
```
Comment (gmarull) [inline dts/bindings/pinctrl/telink,b91-pinctrl.yaml:30]: nit: `Pin functions`. Can description be improved? For example, how do pin function depend on this value?

File: dts/bindings/pinctrl/telink,b91-pinctrl.yaml [hunk 2]
```diff
@@ -0,0 +1,41 @@
+# Copyright (c) 2022, Telink Semiconductor
+# SPDX-License-Identifier: Apache-2.0
+
+description: |
+    Telink B91 Pin controller node.
+    Based on pincfg-node.yaml binding.
```
Comment (gmarull) [inline dts/bindings/pinctrl/telink,b91-pinctrl.yaml:6]: Please extend this description to detail pinctrl usage on this platform. See for example https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/pinctrl/gd%2Cgd32-pinctrl-af.yaml or https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/bindings/pinctrl/renesas%2Crcar-pfc.yaml

File: dts/bindings/pinctrl/telink,b91-pinctrl.yaml [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+child-binding:
+    description: |
+        This binding gives a base representation of the Telink B91 pins configration.
+
+    properties:
+        pinmux:
+          required: true
+          type: int
+          description: |
+              Telink B91 pin's configuration (port, pin, function, pull-up/down).
```
Comment (gmarull) [inline dts/bindings/pinctrl/telink,b91-pinctrl.yaml:41]: `pinmux` field does not encode stuff like pull-up/down

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts [hunk 4]
```diff
@@ -7,6 +7,7 @@
 /dts-v1/;
 
 #include <telink_b91.dtsi>
+#include "tlsr9518adk80d_pinctrl.dtsi"
```
Comment (gmarull) [inline boards/riscv/tlsr9518adk80d/tlsr9518adk80d.dts:10]: nit: `tlsr9518adk80d-pinctrl.dtsi`

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d_pinctrl.dtsi [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+ * Copyright (c) 2022 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <dt-bindings/pinctrl/b91-pinctrl.h>
+
+&pinctrl {
+	/* Enable PINCTRL driver */
+	status = "okay";
```
Comment (gmarull) [inline boards/riscv/tlsr9518adk80d/tlsr9518adk80d_pinctrl.dtsi:11]: note: if all your boards depend in practice on `pinctrl` just enable this node at the SoC dts file.

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d_pinctrl.dtsi [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+
+	uart0_rx_pa4: uart0_rx_pa4 {
+		pinmux = <B91_PINMUX_SET(B91_PORT_A, B91_PIN_4, B91_FUNC_B)>;
+	};
+	uart0_rx_pb3: uart0_rx_pb3 {
+		pinmux = <B91_PINMUX_SET(B91_PORT_B, B91_PIN_3, B91_FUNC_C)>;
+	};
+	uart0_rx_pd3: uart0_rx_pd3 {
+		pinmux = <B91_PINMUX_SET(B91_PORT_D, B91_PIN_3, B91_FUNC_A)>;
+	};
```
Comment (gmarull) [inline boards/riscv/tlsr9518adk80d/tlsr9518adk80d_pinctrl.dtsi:38]: how is it possible that `uart0` `tx`/`rx` have definitions for 3 different pins? If you want to provide all valid combinations, as for example STM32 does, please create a file with valid pin combinations per SoC package instead. You may also want to check the grouping approach, that is, provide something like this instead:

```c
/* provided, may be autogenerated */
#define UART0_RX_PD3 B91_PINMUX_SET(B91_PORT_D, B91_PIN_3, B91_FUNC_A)
```

[GD32](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/gd32f450i_eval/gd32f450i_eval-pinctrl.dtsi) or [RPi Pico](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/rpi_pico/rpi_pico-pinctrl.dtsi) use this model.

I'd also suggest thinking about how you're going to support pin low-power states, since you may find that one model fits better than the other.

File: drivers/pinctrl/Kconfig.b91 [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2022 Telink Semiconductor
+# SPDX-License-Identifier: Apache-2.0
+
+config PINCTRL_TELINK_B91
+	bool "Telink B91 pin controller driver"
+	depends on SOC_RISCV_TELINK_B91
+	default y
```
Comment (gmarull) [inline drivers/pinctrl/Kconfig.b91:7]: please default based on dt_compat_status, you can check some other drivers.

File: drivers/pinctrl/pinctrl_b91.c [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+{
+	uint8_t bit = pin & 0xff;
+
+	reg_gpio_en(pin) &= ~bit;
+}
+
+/* Get function value bits start position (offset) */
+static inline int pinctrl_b91_get_func_offset(uint32_t pin, uint8_t *offset)
+{
+	switch ((pin & 0x0fu) != 0u ? pin & 0x0fu : (pin & 0xf0u) >> 4u) {
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_b91.c:92]: can this be moved out from `switch` for readability?

File: drivers/pinctrl/pinctrl_b91.c [hunk 9]
```diff
@@ -0,0 +1,183 @@
...
+		mask = 0xfc;
+	} else if (pin & 0x22) {
+		val = val << 2;
+		mask = 0xf3;
+	} else if (pin & 0x44) {
+		val = val << 4;
+		mask = 0xcf;
+	} else if (pin & 0x88) {
+		val = val << 6;
+		mask = 0x3f;
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_b91.c:130]: Magic numbers here, please define values.

File: drivers/pinmux/pinmux_b91.c [hunk 10]
```diff
@@ -4,7 +4,7 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#define DT_DRV_COMPAT telink_b91_pinmux
+#define DT_DRV_COMPAT telink_b91_pinctrl
```
Comment (gmarull) [inline drivers/pinmux/pinmux_b91.c:7]: shouldn't this driver just be removed?

File: scripts/west_commands/runners/spi_burn.py [hunk 11]
```diff
@@ -59,28 +59,44 @@ def do_run(self, command, **kwargs):
         else:
             self.logger.error(f'{command} not supported!')
 
-    def _flash(self):
+    def start_iceman(self):
```
Comment (gmarull) [inline scripts/west_commands/runners/spi_burn.py:62]: I'd suggest to move this changes to another PR, since they're completely independent (or looks like they're)

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig [hunk 12]
```diff
@@ -7,7 +7,7 @@ CONFIG_BOARD_TLSR9518ADK80D=y
 CONFIG_PLIC=y
 CONFIG_RISCV_MACHINE_TIMER=y
 CONFIG_SYS_CLOCK_TICKS_PER_SEC=1000
-CONFIG_PINMUX=y
+CONFIG_PINCTRL=y
```
Comment (gmarull) [inline boards/riscv/tlsr9518adk80d/tlsr9518adk80d_defconfig:10]: consider defaulting at SoC level, since pinctrl is in practice always required. See e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/arm/gigadevice/Kconfig.defconfig#L8

**Review by gmarull — CHANGES_REQUESTED** [review]
Thanks for this PR! Please check my comments.

File: dts/bindings/pinctrl/telink,b91-pinctrl.yaml (hunk 1 above)
Comment (yurvyn) [reply dts/bindings/pinctrl/telink,b91-pinctrl.yaml]: Done

File: dts/bindings/pinctrl/telink,b91-pinctrl.yaml (hunk 3 above)
Comment (yurvyn) [reply dts/bindings/pinctrl/telink,b91-pinctrl.yaml]: Fixed

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d_pinctrl.dtsi (hunk 5 above)
Comment (yurvyn) [reply boards/riscv/tlsr9518adk80d/tlsr9518adk80d_pinctrl.dtsi]: Done

File: boards/riscv/tlsr9518adk80d/tlsr9518adk80d_pinctrl.dtsi (hunk 6 above)
Comment (yurvyn) [reply boards/riscv/tlsr9518adk80d/tlsr9518adk80d_pinctrl.dtsi]: Fixed. Left only the definitions which are used by the board.

File: drivers/pinctrl/Kconfig.b91 (hunk 7 above)
Comment (yurvyn) [reply drivers/pinctrl/Kconfig.b91]: Done

File: drivers/pinctrl/pinctrl_b91.c (hunk 8 above)
Comment (yurvyn) [reply drivers/pinctrl/pinctrl_b91.c]: Fixed

File: drivers/pinctrl/pinctrl_b91.c (hunk 9 above)
Comment (yurvyn) [reply drivers/pinctrl/pinctrl_b91.c]: Fixed

File: drivers/pinmux/pinmux_b91.c (hunk 10 above)
Comment (yurvyn) [reply drivers/pinmux/pinmux_b91.c]: Fixed, pinmux is removed.

File: scripts/west_commands/runners/spi_burn.py (hunk 11 above)
Comment (yurvyn) [reply scripts/west_commands/runners/spi_burn.py]: Agree, fixed.

File: soc/riscv/riscv-privilege/telink_b91/pinctrl_soc.h [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+ * Copyright (c) 2022 Telink Semiconductor
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef SOC_RISCV_TELINK_B91_PINCTRL_SOC_H
+#define SOC_RISCV_TELINK_B91_PINCTRL_SOC_H
+
+#include <zephyr/types.h>
+#include <dt-bindings/pinctrl/b91-pinctrl.h>
```
Comment (gmarull) [inline soc/riscv/riscv-privilege/telink_b91/pinctrl_soc.h:11]: nit:
```
#include <stdint.h>
#include <devicetree.h>
#include <dt-bindings/pinctrl/b91-pinctrl.h>
```

File: drivers/pinctrl/pinctrl_b91.c [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	return 0;
+}
+
+/* Set pin's function */
+static int pinctrl_configure_pin(const pinctrl_soc_pin_t *pinctrl)
+{
+	int status = 0;
+	uint8_t mask = 0;
```
Comment (gmarull) [inline drivers/pinctrl/pinctrl_b91.c:129]: nit: status/mask initialization is not required

File: samples/net/openthread/coprocessor/boards/tlsr9518adk80d.overlay [hunk 15]
```diff
@@ -14,11 +14,13 @@
 &uart0 {
```
Comment (gmarull) [inline samples/net/openthread/coprocessor/boards/tlsr9518adk80d.overlay:14]: nit: please update https://github.com/zephyrproject-rtos/zephyr/commit/122bf96ff158d1f22439e6b54908580ee2293ffe commit title (overflows).

**Review by gmarull — APPROVED** [review]
Overall LGTM, no more objections. Would be nice to add a DT parsing test, check for example https://github.com/zephyrproject-rtos/zephyr/tree/main/tests/drivers/pinctrl/gd32

File: samples/net/openthread/coprocessor/boards/tlsr9518adk80d.overlay (hunk 15 above)
Comment (yurvyn) [reply samples/net/openthread/coprocessor/boards/tlsr9518adk80d.overlay:14]: Done

**Comment by yurvyn** [comment]
@kgugala

One more approval is required.
Could you please approve the PR?

**Review by mbolivar-nordic — APPROVED** [review]

**Comment by yurvyn** [comment]
@carlescufi, @nashif
Could you please merge this PR?
